src/v/datalake/translation/scheduling_policies.cc

```diff
@@ -63,4 +63,258 @@ ss::future<> simple_fcfs_scheduling_policy::on_resource_exhaustion(
+            candidates.push_back(
+              {.id = id,
+               .group = translator_group::about_to_expire,
+               .weight = -duration_to_expire.count()});
+        } else {
+            auto total_time = clock::now() - translator.start_time();
+            auto total_running_time = translator.total_running_time();
+            if (total_running_time < minimum_allotment_coeff * total_time) {
+                // within all unfulfilled quota, we want to order order them
+                // by least alloted time first .
+                candidates.push_back(
+                  {.id = id,
+                   .group = translator_group::unfulfilled_quota,
+                   .weight = -total_running_time.count()});
+            } else {
+                // todo: perhaps we can order by pending_bytes_to_translate
+                candidates.push_back(
+                  {.id = id,
+                   .group = translator_group::other,
+                   .weight = random_generators::get_int<long>()});
+            }
+        }
+        const auto& back = candidates.back();
+        vlog(
+          datalake_log.trace,
+          "scheduling candidate: {},  group: {}, weight: {}",
+          translator,
+          back.group,
+          back.weight);
+    }
+
+    co_await ss::maybe_yield();
+
+    // Reactor friendly sort using the comparator
+    std::priority_queue<entry, chunked_vector<entry>, decltype(max_heap_cmp)>
+      prioritized(max_heap_cmp);
+    for (auto& entry : candidates) {
+        executor.as.check();
+        auto it = executor.translators.find(entry.id);
+        if (
+          it == executor.translators.end()
+          || !it->second._waiting_hook.is_linked()) {
+            continue;
+        }
+        prioritized.push(entry);
+        co_await ss::maybe_yield();
+    }
+
+    executor.as.check();
+
+    while (!prioritized.empty()) {
+        auto& entry = prioritized.top();
+        auto it = executor.translators.find(entry.id);
+        if (
+          it != executor.translators.end()
+          && it->second._waiting_hook.is_linked()) {
+            vlog(
+              datalake_log.trace,
+              "[{}] chosen translator to run, group: {}, weight: {}",
+              it->second,
+              entry.group,
+              entry.weight);
+            executor.start_translation(it->second, _translation_time_quota);
+            co_return;
+        }
+        prioritized.pop();
+        co_await ss::maybe_yield();
+    }
+}
+
+ss::future<> fair_scheduling_policy::on_resource_exhaustion(
+  executor& executor, const reservations_tracker& mem_tracker) {
+    if (!mem_tracker.memory_exhausted() || executor.running.empty()) {
+        co_return;
+    }
+    // stop the translator with highest memory usage first.
+    // The sort is stable so it is naturally ordered based on enqueue time
+    // incase of ties. note: the size of this list is super small (low single
```

Comment:
But once we reorder them by bytes reserved, isn't the order no longer strictly enqueue time?

Seems less surprising and maybe cheaper to iterate through? (though i get that this is a small list, :shrug:)

Reply:
As I understood it, the stability of intrusive::list::sort should preserve the enqueue time ordering between two equivalent elements. I agree it's sort of surprising though; I had to convince myself of it again just now.

Reply:
what Oren said.. if two translators have same memory, they will be ordered by their enqueue time.

Reply:
Ah is that a property of intrusive list specifically? I would've thought a stable sort would only preserve the order of right before the sort. So if this list were previously sorted by memory as of the last time on_resource_exhaustion() were called, that would be order in which we break ties.

Reply:
Ah, I misread your original comment,. you are talking about ordering across multiple  on_resource_exhaustion() calls, you are right, I can fix the comment.